src/lib.rs

```diff
@@ -443,9 +457,26 @@ impl<S> StreamUnordered<S> {
             unsafe { *(*task).len_all.get() }
         };
 
-        IterPinMut {
+        IterPinMutWithToken {
+            task,
+            len,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Returns an immutable iterator that allows getting a reference to each stream in the set.
+    pub fn iter_with_token(&self) -> IterWithToken<'_, S> {
+        let task = unsafe { *self.head_all.as_ptr() };
```

Comment:
This function was implemented by me but derived from the impl of `StreamUnordered::iter_pin_mut_with_token`.

I have a feeling this first line is unsafe due to the fact that the method takes `&self` given the comments in the equivalent line within `iter_pin_mut_with_token`, but I am also very outta my depth in this level of unsafe so a full review of this function is in order.

All other functions remain identical minus some internal rearranging.

Reply:
Ah, so, for this you're going to want to use the `atomic_load_head_and_len_all` helper just below, which takes care of _exactly_ reading out the task pointer and length for you based on a `&self`. That should in turn let you make this function entirely safe :+1: